bare_metal_billing/billing.py

```diff
@@ -88,16 +88,29 @@ def _get_su_type(lease_info: models.BMNodeUsage):
 
 
 def _get_running_time(
-    lease_info: models.BMNodeUsage, start_time: datetime, end_time: datetime
-):
+    lease_info: models.BMNodeUsage,
+    start_time: datetime,
+    end_time: datetime,
+    excluded_intervals: list[tuple[datetime, datetime]] = None,
+) -> int:
     start_time = _clamp_time(lease_info.start_time, start_time, end_time)
     end_time = (
         end_time
         if lease_info.expire_time
         is None  # Assumes lease is still running if no expire time given
         else _clamp_time(lease_info.expire_time, start_time, end_time)
     )
-    return math.ceil((end_time - start_time).total_seconds() / 3600)
+
+    total_interval_duration = (end_time - start_time).total_seconds()
+    if excluded_intervals:
+        for e_interval_start, e_interval_end in excluded_intervals:
+            e_interval_start = _clamp_time(e_interval_start, start_time, end_time)
+            e_interval_end = _clamp_time(e_interval_end, start_time, end_time)
+            total_interval_duration -= (
+                e_interval_end - e_interval_start
+            ).total_seconds()
+
+    return math.ceil(total_interval_duration / 3600)
```

Comment:
I know you are rejecting overlapping time ranges and are also clamping the times so that exclusions fall between the runtime, but out of abundance of caution I'd still make sure this never becomes negative. Easy way to do would be something like
```
return math.ceil(max(0, total_interval_duration) /3600)